manager/integration/tests/test_ha.py

```diff
@@ -2675,15 +2775,87 @@ def test_engine_image_not_fully_deployed_perform_volume_operations():
        returns error
     4. Verify that we can attach to another node, take snapshot, take a backup,
        expand, then detach vol-1
-    5. Verify that vol-2 cannot be attached to any nodes because one of
-       its replicas is sitting on the node-1 which doesn't have the
-       engine image. The attach API call returns error
+    5. Verify that vol-2 cannot be attached to tainted nodes. The attach API
+       call returns error
+    6. Verify that vol-2 can attach to non-tainted node with degrade status
     """
-    pass
+    volume1, volume2, tainted_node_id = \
+        prepare_engine_not_fully_deployed_environment_with_volumes(client,
+                                                                   core_api)
 
+    volume1 = client.by_id_volume(volume1.name)
+    volume1 = wait_for_volume_healthy(client, volume1.name)
 
-@pytest.mark.skip(reason="TODO") # NOQA
-def test_engine_image_not_fully_deployed_perform_engine_upgrade():
+    # TODO: write data into volume1.
+    # Did not do data write because volume is not attached to self host node
+
+    # High chance get error "cannot get engine client" first time take snapshot
+    for i in range(RETRY_COUNTS_SHORT):
+        try:
+            snap1 = create_snapshot(client, volume1.name)
+            break
+        except Exception:
+            time.sleep(RETRY_INTERVAL_LONG)
+            continue
+
+    snapshots = volume1.snapshotList()
+    snapMap = {}
+    for snap in snapshots:
+        snapMap[snap.name] = snap
+
+    assert snapMap[snap1.name].name == snap1.name
+
+    volume1.detach(hostId="")
+    volume1 = common.wait_for_volume_detached(client, volume1.name)
+
+    can_not_attach = False
+    try:
+        volume1.attach(hostId=tainted_node_id)
+    except Exception as e:
+        print(e)
+        can_not_attach = True
+
+    assert can_not_attach
+
+    volume1.attach(hostId=get_self_host_id())
+    volume1 = wait_for_volume_healthy(client, volume1.name)
+    snap2_data = write_volume_random_data(volume1)
+    snap2 = create_snapshot(client, volume1.name)
+    snapshots = volume1.snapshotList()
+    snapMap = {}
+    for snap in snapshots:
+        snapMap[snap.name] = snap
+
+    assert snapMap[snap2.name].name == snap2.name
+    check_volume_data(volume1, snap2_data)
+
+    # TODO: For test backup in engine image not full deployed cluser,
```

Comment:
It is okay add code for backup now and let the test fail till https://github.com/longhorn/longhorn/issues/5248 resolved.
We can link the test name [here](https://github.com/longhorn/longhorn/issues/5248), that way it will be easier to track it.

Reply:
Hi @khushboo-rancher, updated. Also I removed the comment because https://github.com/longhorn/longhorn/issues/5248 is ready for test and test case passed in my environment